### PR TITLE
Convert repo-versions tests to pytest

### PR DESCRIPTION
**Pull request by gerrod3**
fixes: #823
Required PR: https://github.com/pulp/pulpcore/pull/3329

### CODE REVIEW

**Comment by gerrod3** [comment]
Testing results:

Test - Old timer, New timer
AddRemoveContentTestCase 12 - 8
SyncChangeRepoVersionTestCase 8 - removed, wasn't testing anything new
AddRemoveRepoVersionTestCase 28 - 13
test_squash_repo_version 13 - 13, didn't change this test
ContentImmutableRepoVersionTestCase 6 - 7
FilterRepoVersionTestCase 2:36 - 11 (This test had eight 10 second sleep statements before)
CreatedResourcesTaskTestCase 6 - 5
CreateRepoBaseVersionTestCase 27 - 10
UpdateRepoVersionTestCase 7 - merged into ContentImmutableRepoVersion
FilterArtifactsTestCase 23 - 8
DeleteRepoVersionResourcesTestCase 16 - 8
ClearAllUnitsRepoVersionTestCase 29 - 10
BaseVersionTestCase 8 - removed, wasn't really testing anything
RepoVersionRetentionTestCase 40 - 13
ContentInRepositoryVersionViewTestCase 8 - 6

Old total: 387 secs
New total: 120 secs
Roughly 3x faster, although all tests are now parallelized (these results are from serial running).

File: pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py [hunk 1]
```diff
@@ -1,77 +1,53 @@
...
+def file_populate_pulp(
+    file_repo_api_client,
+    file_content_api_client,
+    file_fixture_gen_remote_ssl,
+    range_header_manifest_path,
+    gen_object_with_cleanup,
+    orphans_cleanup_api_client,
+):
+    """Fixture to populate Pulp with 8 content units."""
+    repo = gen_object_with_cleanup(file_repo_api_client, gen_repo())
```
Comment (dkliban) [inline pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py:28]: why not use the `file_repo` fixture?

File: pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py [hunk 2]
```diff
@@ -522,858 +347,516 @@ def test_squash_repo_version(
...
-
-        repo = self.client.post(FILE_REPO_PATH, gen_repo())
-        self.addCleanup(self.client.delete, repo["pulp_href"])
+@pytest.mark.parallel
+def test_filter_repo_version(
+    file_repo_api_client, file_repo_ver_api_client, file_populate_pulp, gen_object_with_cleanup
+):
+    """Test whether repository versions can be filtered."""
+    # Setup 8 content units in Pulp to populate test repository with
+    repo = gen_object_with_cleanup(file_repo_api_client, gen_repo())
```
Comment (dkliban) [inline pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py:404]: Seems like you could use the `file_repo` fixture here also.

File: pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py [hunk 3]
```diff
@@ -522,858 +347,516 @@ def test_squash_repo_version(
...
+    for params, num_results in (
+        ({"number": numbers[0]}, 1),
+        ({"number__gt": numbers[0], "number__lt": numbers[-1]}, len(numbers) - 2),
+        ({"number__gte": numbers[0], "number__lte": numbers[-1]}, len(numbers)),
+        ({"number__range": numbers[0:2]}, 2),
+    ):
+        results = file_repo_ver_api_client.list(repo.pulp_href, **params)
+        assert results.count == num_results, params
+
+    # Delete a repository version and filter by its number
```
Comment (dkliban) [inline pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py:473]: Extra indentation here.

File: pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py [hunk 4]
```diff
@@ -522,858 +347,516 @@ def test_squash_repo_version(
...
+    assert results.count == 0
 
-        # create repo version
-        sync(self.cfg, remote, repo)
-        repo = self.client.get(repo["pulp_href"])
 
-        self.assert_patch(repo)
-        self.assert_put(repo)
+@pytest.mark.parallel
+def test_created_resources_task(
```
Comment (dkliban) [inline pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py:480]: This test doesn't seem to be needed. We could make this assertion in any of the other tests.

File: pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py [hunk 5]
```diff
@@ -80,318 +56,167 @@ class AddRemoveContentTestCase(unittest.TestCase):
...
+    file_content_api_client,
+    file_populate_pulp,
+    gen_object_with_cleanup,
+):
+    """Create and delete repository versions."""
+    # Setup 8 content units in Pulp to populate test repository with
+    contents = file_populate_pulp
+
+    # Test trying to delete version 0 on new repository
+    repo = gen_object_with_cleanup(file_repo_api_client, gen_repo())
```
[the remark this answers is not in the document]
Comment (gerrod3) [reply pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py:154]: You are right, I factored out shared code into the `file_populate_pulp` fixture and forgot to switch over to `file_repo`

File: pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py [hunk 6]
```diff
@@ -522,858 +347,516 @@ def test_squash_repo_version(
...
-        }
-        content2 = self.client.using_handler(api.task_handler).post(
-            FILE_CONTENT_PATH, second_content_attrs
-        )
-        modify_repo(self.cfg, repo, add_units=[content2])
-        repo = self.client.get(repo["pulp_href"])
-
-        self.assertEqual(get_content(repo)[FILE_CONTENT_NAME][0], content2)
+    pub_body = {"repository": repo.pulp_href}
+    publication = gen_object_with_cleanup(file_pub_api_client, pub_body)
```
Comment (dkliban) [inline pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py:646]: You could use a `file_repo_with_auto_publish` fixture for this part of the test. 

https://github.com/pulp/pulp_file/blob/main/pulp_file/tests/functional/api/test_auto_publish.py#L22

**Review by dkliban — CHANGES_REQUESTED** [review]
The required changes are with regard to broader use of `file_repo` fixture. I would like to also see the use of the `file_repo_with_auto_publish` where you need publications to be created. 

I believe we could combine most of these tests into one long test. However, it's up to you if you want to try to do that.

File: pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py (hunk 4 above)
Comment (gerrod3) [reply pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py:480]: Removed and placed assertion in the first test.

File: pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py (hunk 3 above)
Comment (gerrod3) [reply pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py:473]: I'm confused, this seems correct to me. The delete is a separate part of the test, it is not apart of the for loop.

File: pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py (hunk 6 above)
Comment (gerrod3) [reply pulp_file/tests/functional/api/from_pulpcore/test_repo_versions.py:646]: I think using less complex features makes the test easier to understand and more confined to what we want to test.

**Review by dkliban — APPROVED** [review]

**Pull request merged**